File: packages/pricing/rokkam_pricing/engine.py
"""Rokkam pricing engine (CLAUDE.md §6).

final = base_price - Σ deductions + Σ bonuses, floored at max(base × 5%, ₹300).
Bonuses are stored as negative deduction values (sign convention shared with the
seed files and the TypeScript mirror in apps/web/src/lib/quote.ts — keep both in
sync; the seed JSON files are the single source of rule data).
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

SCRAP_FLOOR_PCT = 0.05
SCRAP_FLOOR_MIN_INR = 300


@dataclass(frozen=True)
class LedgerLine:
    question_id: str
    option_id: str
    label: str
    amount_inr: int  # signed: negative = deduction, positive = bonus


@dataclass(frozen=True)
class QuoteResult:
    status: str  # "ok" | "declined"
    base_price_inr: int
    ledger: tuple[LedgerLine, ...] = field(default_factory=tuple)
    final_price_inr: int = 0
    floored_at: int | None = None


def _round_half_up(x: float) -> int:
    # Match JS Math.round (half away from zero for positives); Python's round()
    # is banker's rounding and would drift from the web ledger by ₹1.
    return int(x + 0.5)


def scrap_floor(base_inr: int) -> int:
    return max(_round_half_up(base_inr * SCRAP_FLOOR_PCT), SCRAP_FLOOR_MIN_INR)


def _amount_for(base_inr: int, dtype: str, value: float) -> int:
    magnitude = _round_half_up(base_inr * abs(value) / 100) if dtype == "pct" else _round_half_up(abs(value))
    return -magnitude if value > 0 else magnitude


def compute_quote(matrix: dict[str, Any], base_inr: int, answers: dict[str, Any]) -> QuoteResult:
    """Pure quote computation over a deduction matrix (seeds/pricing/*_deductions.json shape)."""
    ledger: list[LedgerLine] = []

    for section in matrix["sections"]:
        for question in section["questions"]:
            answer = answers.get(question["id"])
            if answer is None:
                continue
            chosen = answer if isinstance(answer, list) else [answer]
            for option_id in chosen:
                option = next((o for o in question["options"] if o["id"] == option_id), None)
                if option is None:
                    continue
                if option.get("kills_deal"):
                    return QuoteResult(status="declined", base_price_inr=base_inr)
                deduction = option.get("deduction")
                if deduction:
                    ledger.append(
                        LedgerLine(
                            question_id=question["id"],
                            option_id=option["id"],
                            label=option["label_en"],
                            amount_inr=_amount_for(base_inr, deduction["type"], deduction["value"]),
                        )
                    )

    raw = base_inr + sum(line.amount_inr for line in ledger)
    floor = scrap_floor(base_inr)
    floored = raw < floor
    return QuoteResult(
        status="ok",
        base_price_inr=base_inr,
        ledger=tuple(ledger),
        final_price_inr=floor if floored else raw,
        floored_at=floor if floored else None,
    )


def composed_base(entry: dict[str, Any], selection: dict[str, str]) -> int:
    """Composed-mode (laptop) base: base_config buyback + selected axis modifiers."""
    total = int(entry["base"])
    for axis, options in entry["axes"].items():
        chosen = selection.get(axis)
        if chosen is not None and chosen in options:
            total += int(options[chosen])
    return total

File: apps/seller_app/lib/data/engine.dart
// Dart port of the pricing engine (packages/pricing, mirrored in
// apps/web/src/lib/quote.ts). Keep behaviour in sync across all three.

import 'dart:math';

import 'models.dart';

const quoteLockDays = 7;

/// Scrap-value floor: max(base × 5%, ₹300) — CLAUDE.md §6.
int scrapFloor(int baseInr) => max((baseInr * 0.05).round(), 300);

int _amountFor(int baseInr, String type, num value) {
  // Sign convention: value > 0 is a deduction, value < 0 a bonus.
  final magnitude = type == 'pct' ? ((baseInr * value.abs()) / 100).round() : value.abs().toInt();
  return value > 0 ? -magnitude : magnitude;
}

/// Pure quote computation: base − Σ deductions + Σ bonuses, floored at scrap value.
QuoteResult computeQuote(DeductionMatrix matrix, int baseInr, Answers answers) {
  final ledger = <LedgerLine>[];

  for (final section in matrix.sections) {
    for (final q in section.questions) {
      final answer = answers[q.id];
      if (answer == null) continue;
      final chosen = answer is List ? answer.cast<String>() : [answer as String];
      for (final optionId in chosen) {
        final matches = q.options.where((o) => o.id == optionId);
        if (matches.isEmpty) continue;
        final opt = matches.first;
        if (opt.killsDeal) {
          return QuoteResult(status: 'declined', basePriceInr: baseInr, ledger: const [], finalPriceInr: 0);
        }
        final d = opt.deduction;
        if (d != null) {
          ledger.add(LedgerLine(q.id, opt.id, opt.labelEn, _amountFor(baseInr, d.type, d.value)));
        }
      }
    }
  }

  final raw = baseInr + ledger.fold<int>(0, (sum, line) => sum + line.amountInr);
  final floor = scrapFloor(baseInr);
  final floored = raw < floor;
  return QuoteResult(
    status: 'ok',
    basePriceInr: baseInr,
    ledger: ledger,
    finalPriceInr: floored ? floor : raw,
    flooredAt: floored ? floor : null,
  );
}

/// Effective base for a composed-mode device (laptops): base_config buyback
/// plus the INR modifier of each selected axis label.
int composedBase(ComposedPriceEntry entry, Map<String, String> selection) {
  var total = entry.base;
  for (final MapEntry(key: axis, value: options) in entry.axes.entries) {
    final chosen = selection[axis];
    if (chosen != null && options[chosen] != null) total += options[chosen]!;
  }
  return total;
}

/// Demo quote code, e.g. RKM-7F3KQ2 (the server issues real codes).
String demoQuoteCode() {
  const alphabet = '23456789ABCDEFGHJKMNPQRSTUVWXYZ';
  final rng = Random();
  return 'RKM-${List.generate(6, (_) => alphabet[rng.nextInt(alphabet.length)]).join()}';
}

/// ₹12,345 (en-IN grouping: last 3, then 2s).
String formatInr(int amount) {
  final sign = amount < 0 ? '−' : '';
  var digits = amount.abs().toString();
  final parts = <String>[];
  if (digits.length > 3) {
    parts.add(digits.substring(digits.length - 3));
    digits = digits.substring(0, digits.length - 3);
    while (digits.length > 2) {
      parts.add(digits.substring(digits.length - 2));
      digits = digits.substring(0, digits.length - 2);
    }
    parts.add(digits);
    digits = parts.reversed.join(',');
  }
  return '$sign₹$digits';
}

/// "− ₹1,470" / "+ ₹250" — signed ledger amounts.
String formatSigned(int amount) => amount < 0 ? '− ${formatInr(-amount)}' : '+ ${formatInr(amount)}';
